qualtran/bloqs/rotations/phase_gradient.py

```diff
@@ -183,21 +197,28 @@ def scaled_val(self, x: int) -> int:
         x_fxp = _fxp(x / 2**x_width, x_width).like(_fxp(0, self.phase_bitsize)).astype(float)
         return int(x_fxp.astype(float) * 2**self.phase_bitsize)
 
-    def apply(self, x: int, phase_grad: int) -> Union[int, Iterable[int]]:
-        out = self.on_classical_vals(x=x, phase_grad=phase_grad)
-        return out['x'], out['phase_grad']
+    def apply(self, ctrl: int, x: int, phase_grad: int) -> Union[int, Iterable[int]]:
+        out = self.on_classical_vals(ctrl=ctrl, x=x, phase_grad=phase_grad)
+        return out['ctrl'], out['x'], out['phase_grad']
 
-    def on_classical_vals(self, x: int, phase_grad: int) -> Dict[str, 'ClassicalValT']:
-        phase_grad_out = (phase_grad + self.sign * self.scaled_val(x)) % (2**self.phase_bitsize)
-        return {'x': x, 'phase_grad': phase_grad_out}
+    def on_classical_vals(self, ctrl: int, x: int, phase_grad: int) -> Dict[str, 'ClassicalValT']:
+        phase_grad_out = (phase_grad + self.sign * ctrl * self.scaled_val(x)) % (
+            2**self.phase_bitsize
+        )
+        return {'ctrl': ctrl, 'x': x, 'phase_grad': phase_grad_out}
 
     def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
-        num_toffoli = self.phase_bitsize - 2
-        return {(Toffoli(), num_toffoli)}
+        if self.controlled == False:
+            num_toffoli = self.phase_bitsize - 2
+            return {(Toffoli(), num_toffoli)}
+        return {(TGate(), 8 * self.phase_bitsize)}
```

Comment:
If `num_toffoli = self.phase_bitsize-2`, the call graph should decompose into `num_toffoli` $\text{And}$ and $\text{And}^\dagger$'s for the non-controlled case and `2 * num_toffoli` of each for the controlled case. 


The fact that we need twice the number of $\text{And}$ / $\text{And}^\dagger$ comes from https://arxiv.org/pdf/1709.06648.pdf (see primitives for non-controlled additions and controlled additions below)


<img width="852" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/25021d49-dfbf-4b64-a0e2-78add15f480b">


<img width="852" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/c92fb58a-5fe8-4c28-9580-2376fd570175">


And the fact that we need `self.phase_bitsize - 2` instead of `self.phase_bitsize - 1` comes from the optimization described in Fig 13 of https://arxiv.org/pdf/2007.07391.pdf where they get rid of the last carry bit utilizing the fact that target register is a phase gradient register